### PR TITLE
DateBox: hide the validation icon when widget size is small to fully display the date (T743726)

### PR DESCRIPTION
**Pull request by ksercs**

### CODE REVIEW

File: js/ui/date_box/ui.date_box.base.js [hunk 1]
```diff
@@ -407,42 +398,32 @@ var DateBox = DropDownEditor.inherit({
...
-            isWidthSet = typeUtils.isDefined(widthOption) || (isRealWidthSet($element) && !this._isSizeUpdatable),
-            pickerType = this._pickerType,
-            // NOTE: no calculateWidth if type is rollers, why?
-            shouldCalculateWidth = pickerType !== PICKER_TYPE.rollers && devices.current().platform === "generic";
-
-        if(!windowUtils.hasWindow() || isWidthSet || !(shouldCalculateWidth && $element.is(":visible"))) {
-            return;
-        }
+    _formatValidationIcon: function(value) {
+        const longestValue = "12-12-1900";
```
Comment (mrjohn0011) [inline js/ui/date_box/ui.date_box.base.js]: Why 12-12-1900 is the longest value?

Comment (ksercs) [reply js/ui/date_box/ui.date_box.base.js]: Fixed: longest value calculates automatically now.

File: js/ui/date_box/ui.date_box.base.js [hunk 2]
```diff
@@ -309,6 +309,7 @@ var DateBox = DropDownEditor.inherit({
 
     _init: function() {
         this._initStrategy();
+        this.$element().addClass("dx-show-invalid-badge");
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.base.js]: Replace class name with a const. It should be defined at the head of file like `DX_AUTO_WIDTH_CLASS`

File: js/ui/date_box/ui.date_box.base.js [hunk 3]
```diff
@@ -445,6 +447,36 @@ var DateBox = DropDownEditor.inherit({
         return width;
     },
 
+    _formatValidationIcon: function(longestValue) {
+        const $input = this._input();
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.base.js]: Do not mix es6 and es5, choose one of them

**Review by DokaRus — COMMENTED** [review]
Does it work well with additional action [buttons](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxDateBox/Configuration/buttons/)?

File: js/ui/date_box/ui.date_box.base.js (hunk 3 above)
Comment (ksercs) [reply js/ui/date_box/ui.date_box.base.js]: I tested it with custom buttons from this demo: https://demos.devexpress.com/ASPNetCore/Demo/Common/CustomTextEditorButtons/. It works.

File: js/ui/date_box/ui.date_box.base.js [hunk 4]
```diff
@@ -407,42 +399,45 @@ var DateBox = DropDownEditor.inherit({
...
-        $element.width(this._calculateWidth(longestValue));
-        this._isSizeUpdatable = true;
-    },
-
-    _calculateWidth: function(value) {
-        var IE_ROUNDING_ERROR = 10;
-        var NATIVE_BUTTONS_WIDTH = 48;
         var $input = this._input();
-        var $longestValueElement = dom.createTextElementHiddenCopy($input, value);
+        var $editorInput = $input.get(0);
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.base.js]:
```suggestion
        var inputElement= $input.get(0);
```

File: js/ui/date_box/ui.date_box.base.js [hunk 5]
```diff
@@ -407,42 +399,45 @@ var DateBox = DropDownEditor.inherit({
...
+        var $dateBox = this.$element();
+
+        if(this._needsValidationIconHiding(longestValue, $dateBox, $input, $editorInput)) {
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.base.js]: Is it show validation icon after update widget's width runtime?

```suggestion
        var shouldHideValidationIcon = this._needsValidationIconHiding(longestValue, $dateBox, $input, $editorInput); 
        
        $dateBox.toggleClass(DX_INVALID_BADGE_CLASS, !shouldHideValidationIcon );
        if(shouldHideValidationIcon ) {
```
and remove L416-L418

File: js/ui/date_box/ui.date_box.base.js [hunk 6]
```diff
@@ -407,42 +399,45 @@ var DateBox = DropDownEditor.inherit({
...
+        var elementWidth = parseFloat(window.getComputedStyle($longestValueElement.get(0)).width);
+        var rightPadding = parseFloat(window.getComputedStyle($editorInput).paddingRight);
+        var leftPadding = parseFloat(window.getComputedStyle($editorInput).paddingLeft);
+        var necessaryWidth = elementWidth + leftPadding + rightPadding;
         $longestValueElement.remove();
 
-        return width;
+        var clearButtonWidth = 0;
+        if(this.option("showClearButton") && $input.val() === "") {
+            var $clearButton = $dateBox.find(".dx-clear-button-area");
```
Comment (DokaRus) [inline js/ui/date_box/ui.date_box.base.js]: Extract class name

File: testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js [hunk 7]
```diff
@@ -1340,6 +1340,14 @@ QUnit.module("widget sizing render", {}, () => {
         assert.ok($element.outerWidth() <= 100, "outer width of the element must be less or equal to a container width");
     });
 
+    QUnit.test("validation icon should hide if container size is too small", assert => {
+        const $element = $("#innerDateBox").dxDateBox({
+            "showClearButton": true
+        });
+
+        assert.notOk($element.hasClass('dx-show-invalid-badge'), "validation icon's hidden");
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/datebox.tests.js:1349]: Then try to increase the container width and check that element has it class

File: js/ui/date_box/ui.date_box.base.js (hunk 5 above)
Comment (ksercs) [reply js/ui/date_box/ui.date_box.base.js]: No. But it's good test case, thanks. I've tried to repair. Please, check it.

**Review by DokaRus — APPROVED** [review]

**Review by mrjohn0011 — APPROVED** [review]

**Pull request merged**